irc/server.go

```diff
@@ -226,9 +229,105 @@ func (server *Server) checkTorLimits() (banned bool, message string) {
 
 // createListener starts a given listener.
 func (server *Server) createListener(addr string, conf listenerConfig, bindMode os.FileMode) (*ListenerWrapper, error) {
-	// make listener
+	if conf.WebSocket {
+		return server.createWSListener(addr, conf)
+	}
+	return server.createNetListener(addr, conf, bindMode)
+}
+
+func (server *Server) isTrusted(ip string) bool {
+	netIP := net.ParseIP(ip)
+	return utils.IPInNets(netIP, server.Config().Server.proxyAllowedFromNets)
+}
+
+func (server *Server) followHTTPForwards(addr string, forwards string) string {
+	if !server.isTrusted(addr) {
+		return addr
+	}
+
+	forwardIPs := strings.Split(forwards, ",")
+
+	// Reverse the slice, to have the inner-most proxy first.
+	for i, j := 0, len(forwardIPs)-1; i < j; i, j = i+1, j-1 {
+		forwardIPs[i], forwardIPs[j] = forwardIPs[j], forwardIPs[i]
+	}
+
+	for i := range forwardIPs {
+		// Using i so that addr points to the last item after the end of the loop.
+		addr = forwardIPs[i]
+
+		if !server.isTrusted(addr) {
+			return addr
+		}
+	}
+
+	// All IPs are trusted? weird. Let's take the last one and call it a day.
+	return addr
+}
+
+// createWSListener starts a given WebSocket listener.
+func (server *Server) createWSListener(addr string, conf listenerConfig) (*ListenerWrapper, error) {
+	var listener net.Listener
+	var err error
+
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		remoteAddr := r.RemoteAddr
+		if header, ok := r.Header["X-Forwarded-For"]; ok {
+			remoteAddr = server.followHTTPForwards(remoteAddr, header[len(header)-1])
+		}
+
+		conn, err := wsUpgrader.Upgrade(w, r, nil)
+		if err != nil {
+			server.logger.Error("internal", "upgrade error", addr, err.Error())
+			return
+		}
+
+		newConn := clientConn{
+			Conn:   WSContainer{conn},
+			Config: conf,
+		}
+
+		server.RunClient(newConn, "")
+	}
+	endpoint := http.Server{
+		Addr:           addr,
+		Handler:        http.HandlerFunc(handler),
+		ReadTimeout:    10 * time.Second,
+		WriteTimeout:   10 * time.Second,
+		MaxHeaderBytes: 1 << 20,
+	}
+	if conf.TLSConfig != nil {
+		listener, err = tls.Listen("tcp", addr, conf.TLSConfig)
+	} else {
+		listener, err = net.Listen("tcp", addr)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// throw our details to the server so we can be modified/killed later
+	wrapper := ListenerWrapper{
+		listener:   listener,
+		httpServer: &endpoint,
+		config:     conf,
+		shouldStop: false,
+	}
+
+	go func() {
+		err := endpoint.Serve(listener)
```

Comment:
I really like the separation of Listen and Serve here. That should make it easy to implement rehash later.